R/messages.R

```diff
@@ -110,3 +110,229 @@ setvar <- function(core, dates, var, values, unit)
+
+    invisible(rslt_tot)
+}
+
+
+
+#' Fetch all the available Hector variables
+#'
+#' This function is similar to fetchvars, except it fetches all available
+#' Hector variables, including exogenous inputs (e.g., emissions & pre-industrial
+#' concentrations), parameters (e.g., alpha, beta, ECS), and output
+#' variables (e.g., concentrations & forcings).
+#'
+#' @param core Hector core object
+#' @param dates ector of dates, optional
+#' @param scenario Scenario name, optional str
+#' @param write If TRUE, write the resulting dataframe to a csv file, optional bool
+#' @param outpath Absolute path of the output csv. Must be given if 'write' is true
+#' @return vars_all    Dataframe containing all Hector variables
+#' @family main user interface functions
+#' @export
+fetchvars_all <- function(core, dates=NULL, scenario=NULL, write=F, outpath=NULL) {
+
+    if (is.null(scenario)) {
+        scenario <- core$name
+    }
+
+    # Load the list of available functions in the hector R library.
+    # Will be passed to helper funcs to match capability strings using regex
+    lib_funcs <- lsf.str("package:hector")
+
+    strt <- startdate(core)
+    end <- getdate(core)
+
+    if (is.null(dates)) {
+        dates <- strt:end
+    } else {
+        valid <- dates >= strt & dates <= end
+        dates <- dates[valid]
+    }
+
+    # Get variables from the helper functions, combine results into one dataframe
+    vars_inpt  <- cum_vars_input(core, dates, lib_funcs)
+    vars_param <- cum_vars_params(core)
+    vars_outpt <- cum_vars_output(core, dates, lib_funcs)
+
+    vars_all <- rbind(vars_inpt, vars_param, vars_outpt)
+
+    # If 'write' is TRUE, reshape the resulting data frame to a wide format
+    # and write to csv
+    if (write) {
+
+        # Pivot the dataframe "wide" so years run along the x axis
+        # Catch warning from reshape()
+        suppressWarnings(vars_all <- reshape(vars_all, direction="wide",
+                                     idvar=c("variable", "units"), timevar="year"))
+
+        # Clean up the column names (value.YYYY --> YYYY)
+        col_names <- colnames(vars_all)[-1:-2]
+        col_names <- sapply(col_names, function(x) sub("value.", "", x), USE.NAMES=F)
+        col_names <- col_names[-length(col_names)]
+
+        col_names <- c("variable", "units", "initial value", col_names)
+
+        num_cols <- length(col_names)
+
+        # Move last column into 3rd col position
+        vars_all <- vars_all[, c(1, 2, num_cols, 4:num_cols-1)]
+
+        # Set the column names for the re-ordered dataframe
+        colnames(vars_all) <- col_names
+
+        # Check if the absolute path of the output csv file is valid
+        if (!is.null(outpath)) {
+            write.table(vars_all, outpath, col.names=col_names, row.names=F, sep=',')
+            cat("Output written to ", outpath, sep="")
+        } else {
+            # Write to a default location in the Hector R library output directory
```

Comment:
I am very leery of writing stuff into R's package directories outside of the package install mechanism.  If nothing else, we don't even know for sure that the user has write privilege there.  It also puts us in the position of having "installed" files that the package doesn't know anything about.  If we need a default output path, we should put it in the current working directory (better still, make specifying the output file the thing that turns on the file output)